voxblox/src/integrator/tsdf_integrator.cc

```diff
@@ -0,0 +1,567 @@
+          base_voxel.color, base_voxel.weight, temp_voxel.second.color,
+          temp_voxel.second.weight);
+    }
+
+    base_voxel.distance =
+        (base_voxel.distance * base_voxel.weight +
+         temp_voxel.second.distance * temp_voxel.second.weight) /
+        new_weight;
+
+    base_voxel.weight = std::min(config_.max_weight, new_weight);
+  }
+}
+
+// Updates tsdf_voxel. Thread safe.
+inline void TsdfIntegratorBase::updateTsdfVoxel(
+    const Point& origin, const Point& point_G, const Point& voxel_center,
+    const Color& color, const float weight, TsdfVoxel* tsdf_voxel) {
+  DCHECK(tsdf_voxel != nullptr);
+
+  const float sdf = computeDistance(origin, point_G, voxel_center);
+
+  float updated_weight = weight;
+  // Compute updated weight in case we use weight dropoff. It's easier here
+  // that in getVoxelWeight as here we have the actual SDF for the voxel
+  // already computed.
+  const FloatingPoint dropoff_epsilon = voxel_size_;
+  if (config_.use_weight_dropoff && sdf < -dropoff_epsilon) {
+    updated_weight = weight * (config_.default_truncation_distance + sdf) /
+                     (config_.default_truncation_distance - dropoff_epsilon);
+    updated_weight = std::max(updated_weight, 0.0f);
+  }
+
+  // Compute the updated weight in case we compensate for sparsity. By
+  // multiplicating the weight of occupied areas (|sdf| < truncation distance)
+  // by a factor, we prevent to easily fade out these areas with the free
+  // space parts of other rays which pass through the corresponding voxels.
+  // This can be useful for creating a TSDF map from sparse sensor data (e.g.
+  // visual features from a SLAM system). By default, this option is disabled.
+  if (config_.use_sparsity_compensation_factor) {
+    if (std::abs(sdf) < config_.default_truncation_distance) {
+      updated_weight *= config_.sparsity_compensation_factor;
+    }
+  }
+
+  // Lookup the mutex that is responsible for this voxel and lock it
+  std::lock_guard<std::mutex> lock(
+      mutexes_.get(getGridIndexFromPoint(point_G, voxel_size_inv_)));
+
+  const float new_weight = tsdf_voxel->weight + updated_weight;
+
+  // it is possible to have weights very close to zero, due to the limited
+  // precision of floating points dividing by this small value can cause nans
+  if (new_weight < kFloatEpsilon) {
+    return;
+  }
+
+  const float new_sdf =
+      (sdf * updated_weight + tsdf_voxel->distance * tsdf_voxel->weight) /
+      new_weight;
+
+  // color blending is expensive only do it close to the surface
+  if (std::abs(sdf) < config_.default_truncation_distance) {
+    tsdf_voxel->color = Color::blendTwoColors(
+        tsdf_voxel->color, tsdf_voxel->weight, color, updated_weight);
+  }
+
+  tsdf_voxel->distance =
+      (new_sdf > 0.0) ? std::min(config_.default_truncation_distance, new_sdf)
+                      : std::max(-config_.default_truncation_distance, new_sdf);
+  tsdf_voxel->weight = std::min(config_.max_weight, new_weight);
```

Comment:
I was mimicking the behavior already present. I am not sure what is best here, will discuss.